### PR TITLE
Add restricted method text

### LINKED ISSUE #1195: Update definition for "restricted" methods
We currently imply that a "restricted" method is a method that requires being passed directly to `wallet_requestPermissions`:
- https://docs.metamask.io/wallet/reference/json-rpc-api/
- https://docs.metamask.io/wallet/how-to/manage-permissions/
- https://docs.metamask.io/wallet/concepts/wallet-api/#restricted-methods

Under this definition, only `eth_accounts` is restricted. We want to modify this definition to include the signature, sending, and encryption methods that are enabled via `eth_accounts`.

Something along the lines of "restricted methods are methods that cannot be called unless you have permission to do so using `wallet_requestPermissions` or `wallet_requestSnaps`." Add a note to this definition stating that acquiring the `eth_accounts` permission includes/enables calling the restricted signing/sending methods.

Also include in the description of each restricted method how to acquire the permission (either via `eth_requestAccounts` or `wallet_requestSnaps` – this is already be done for the restricted Snaps methods).

See https://github.com/MetaMask/api-specs/pull/195 for the full list of restricted methods.

### PR DESCRIPTION
**Pull request by joaniefromtheblock**
Partially closes #1195 

Update restricted methods definition and description.

Docs changes on the following pages :
- https://docs.metamask.io/update-restricted/wallet/concepts/wallet-api/
- https://docs.metamask.io/snaps/how-to/request-permissions/
- https://docs.metamask.io/update-restricted/wallet/how-to/manage-permissions/
- https://github.com/MetaMask/metamask-docs/pull/1217/files

### CODE REVIEW

File: wallet/concepts/wallet-api.md [hunk 1]
```diff
@@ -61,16 +61,19 @@ Make sure to handle errors for every call to
...
 Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
 internally by MetaMask.
 
-Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), the only
-restricted method is [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
-the user's Ethereum accounts.
-More restricted methods will be added in the future.
+Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), there is the
+restricted method [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
+the user's Ethereum accounts. DApps must explicitly request user consent using `wallet_requestPermissions` before using `eth_accounts`.
```
Comment (joaniefromtheblock) [inline wallet/concepts/wallet-api.md:73]: Dapp

File: wallet/concepts/wallet-api.md [hunk 2]
```diff
@@ -61,16 +61,19 @@ Make sure to handle errors for every call to
...
+Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [wallet_requestSnaps](/wallet/reference/wallet_requestSnaps).
+
 Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
 internally by MetaMask.
 
-Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), the only
-restricted method is [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
-the user's Ethereum accounts.
-More restricted methods will be added in the future.
+Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), there is the
```
Comment (joaniefromtheblock) [inline wallet/concepts/wallet-api.md:71]: eth_accounts requires request permissions using ...
you need request permissions for the wallet_ snap / wallet_invokeSnap  using the wallet_request

File: wallet/concepts/wallet-api.md (hunk 1 above)
Comment (joaniefromtheblock) [reply wallet/concepts/wallet-api.md:73]: As well as add a NOTE about eth_requestAccounts

Comment (joaniefromtheblock) [reply wallet/concepts/wallet-api.md:73]: @alexandratran suggestion 

The following methods are restricted:
- eth_accounts - Requires passing eth_accounts as a parameter to wallet_requestPermissions. This unlocks:
  - eth_sendTransaction
  - personal_sign
  - eth_signTypedData_v4
note: we recommend calling eth_requestAccounts, which internally calls wallet_requestPermissions on eth_accounts
- wallet_snap - Requires calling wallet_requestSnap
- wallet_invokeSnap - Requires calling wallet_requestSnap


To call wallet_snap or wallet_invokeSnap on a Snap, you need to first call wallet_requestSnap on that Snap
Example on the bottom of this section: https://docs.metamask.io/snaps/learn/about-snaps/apis/#custom-json-rpc-apis (edited)

File: wallet/concepts/wallet-api.md [hunk 3]
```diff
@@ -61,16 +61,19 @@ Make sure to handle errors for every call to
...
+restricted method [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
+the user's Ethereum accounts. DApps must explicitly request user consent using `wallet_requestPermissions` before using `eth_accounts`.
+
+Granting permission for `eth_accounts` also implicitly grants access to [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction), [`personal_sign`](/wallet/reference/personal_sign), and [`eth_signTypedData`](/wallet/reference/eth_signTypedData). 
+
```
Comment (joaniefromtheblock) [inline wallet/concepts/wallet-api.md:76]: make sure no excessive line breaks

File: wallet/concepts/wallet-api.md [hunk 4]
```diff
@@ -61,16 +61,19 @@ Make sure to handle errors for every call to
...
 
 ### Unrestricted methods
```
Comment (joaniefromtheblock) [inline wallet/concepts/wallet-api.md:78]: remove the might still rely on to succeed since we are including this in the definition now

File: wallet/concepts/wallet-api.md [hunk 5]
```diff
@@ -61,23 +61,44 @@ Make sure to handle errors for every call to
...
+MetaMask implements permissions based on [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255) to enhance security for when users interact with dapps. 
+This requires that dapps obtain user consent before accessing certain features. 
 Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
 internally by MetaMask.
 
-Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), the only
-restricted method is [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
-the user's Ethereum accounts.
-More restricted methods will be added in the future.
+Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [wallet_requestSnaps](/wallet/reference/wallet_requestSnaps).
```
Comment (Montoya) [inline wallet/concepts/wallet-api.md:77]: Missed the \` \` on wallet_requestSnaps

**Review by Montoya — CHANGES_REQUESTED** [review]

File: wallet/concepts/wallet-api.md [hunk 6]
```diff
@@ -23,23 +23,20 @@ This API is specified by [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193), an
...
-alternative wallet detection mechanism to the `window.ethereum` injected provider.
-This alternative mechanism enables dapps to support [wallet interoperability](wallet-interoperability.md)
+alternative discovery mechanism to the `window.ethereum` injected provider.
+This alternative mechanism enables dapps to support [wallet interoperability](../wallet-interoperability)
 by discovering multiple injected wallet providers in a user's browser.
-We recommend [using this mechanism to connect to MetaMask](../how-to/connect/index.md).
-
-You can access the provider API using the selected EIP-6963 provider object.
-Throughout this documentation, we refer to the selected provider using `provider`.
+We recommend using this mechanism for provider discovery.
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:31]: Please revert these changes, which revert merged changes in #1214.

File: wallet/concepts/wallet-api.md [hunk 7]
```diff
@@ -23,23 +23,20 @@ This API is specified by [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193), an
...
 :::
 
 The MetaMask Ethereum provider API contains the following:
 
 - [Properties](../reference/provider-api.md#properties) - The provider contains a property that
   detects if a user has MetaMask installed.
 - [Methods](../reference/provider-api.md#methods) - The provider contains methods that dapps can call.
-  The [`request()`](../reference/provider-api.md#request)
+  The [`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs)
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:39]: Please revert this change, which reverts a merged change in #1214.

File: wallet/concepts/wallet-api.md [hunk 8]
```diff
@@ -49,7 +46,7 @@ methods, and events.
 
 ## JSON-RPC API
 
-MetaMask uses the [`request()`](../reference/provider-api.md#request)
+MetaMask uses the [`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs)
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:49]: Please revert this change, which reverts a merged change in #1214.

File: wallet/concepts/wallet-api.md [hunk 9]
```diff
@@ -58,26 +55,46 @@ The RPC methods are documented in the interactive [JSON-RPC API reference](/wall
 :::note
 All RPC method requests can return errors.
 Make sure to handle errors for every call to
-[`request()`](../reference/provider-api.md#request).
+[`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs).
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:58]: Please revert this change, which reverts a merged change in #1214.

File: wallet/concepts/wallet-api.md [hunk 10]
```diff
@@ -58,26 +55,46 @@ The RPC methods are documented in the interactive [JSON-RPC API reference](/wall
...
+To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts).
+This method automatically obtains permissions for `eth_accounts` through an internal `wallet_requestPermissions` call.
+See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
+:::
+
+- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
+- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
+  
+:::info note
+For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](../../../../snaps/learn/about-snaps/apis.md#custom-json-rpc-apis) page.
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:94]:
```suggestion
For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](/snaps/learn/about-snaps/apis/#custom-json-rpc-apis) page.
```

Sorry for the confusion, but when linking between `wallet` and `snaps`, you need to use absolute links.

**Review by alexandratran — CHANGES_REQUESTED** [review]

File: snaps/how-to/request-permissions.md [hunk 11]
```diff
@@ -12,6 +12,10 @@ to communicate with Snaps.
 
 Snaps and dapps follow the [EIP-2255 wallet permissions specification](https://eips.ethereum.org/EIPS/eip-2255).
 
+  :::info note
+  The [JSON-RPC](/wallet/reference/json-rpc-api) methods [`wallet_snap`](/wallet/reference/wallet_snap) and [`wallet_invokeSnap`](/wallet/reference/wallet_invokesnap) are restricted. To use these restricted methods, you must first obtain permission through [`wallet_requestSnap`](/wallet/reference/wallet_requestSnaps).
```
Comment (Montoya) [inline snaps/how-to/request-permissions.md:16]: Should be [`wallet_requestSnaps`]

**Review by Montoya — CHANGES_REQUESTED** [review]

**Review by Montoya — APPROVED** [review]

File: snaps/how-to/request-permissions.md [hunk 12]
```diff
@@ -12,6 +12,10 @@ to communicate with Snaps.
 
 Snaps and dapps follow the [EIP-2255 wallet permissions specification](https://eips.ethereum.org/EIPS/eip-2255).
 
+  :::info note
+  The [JSON-RPC](/wallet/reference/json-rpc-api) methods [`wallet_snap`](/wallet/reference/wallet_snap) and [`wallet_invokeSnap`](/wallet/reference/wallet_invokesnap) are restricted. To use these restricted methods, you must first obtain permission through [`wallet_requestSnaps`](/wallet/reference/wallet_requestSnaps).
+  :::
+
```
Comment (alexandratran) [inline snaps/how-to/request-permissions.md:18]: This is not needed in this location. There is a "Request permissions from a dapp" section at the bottom of this page that mentions `wallet_requestSnaps`. You can add mention of `wallet_snap` and `wallet_invokeSnap` in that section.

File: wallet/concepts/wallet-api.md [hunk 13]
```diff
@@ -61,23 +61,43 @@ Make sure to handle errors for every call to
 [`request()`](../reference/provider-api.md#request).
 :::
 
+Dapps communicate with MetaMask through JSON-RPC methods. 
+These methods are divided into the following:
+
+- [Restricted methods](#restricted-methods) -  Require user consent for actions that impact assets or data (for example initiating a transaction).
+- [Unrestricted methods](#unrestricted-methods) - Allow dapps to perform basic actions without permission (for example retrieving a public address).
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:68]:
```suggestion
The RPC methods are divided into the following:

- [Restricted methods](#restricted-methods) -  Require user consent for actions that impact assets or data (for example, initiating a transaction).
- [Unrestricted methods](#unrestricted-methods) - Allow dapps to perform basic actions without permission (for example, retrieving a public address).
```

File: wallet/concepts/wallet-api.md [hunk 14]
```diff
@@ -61,23 +61,43 @@ Make sure to handle errors for every call to
...
-Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), the only
-restricted method is [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access
-the user's Ethereum accounts.
-More restricted methods will be added in the future.
+Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [`wallet_requestSnaps`](/wallet/reference/wallet_requestSnaps).
+
+The following methods are restricted:
+
+- [eth_accounts](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
+Granting permissions for `eth_accounts` also provides permissions for the following methods:
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:82]:
```suggestion
- [`eth_accounts`](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
Granting permission for `eth_accounts` also grants permissions for the following methods:
```

File: wallet/concepts/wallet-api.md [hunk 15]
```diff
@@ -61,23 +61,43 @@ Make sure to handle errors for every call to
...
+  - [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction)
+  - [`personal_sign`](/wallet/reference/personal_sign)
+  - [`eth_signTypedData_v4`](/wallet/reference/eth_signTypedData_v4)
+
+:::info note
+To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts).
+This method automatically obtains permissions for `eth_accounts` through an internal `wallet_requestPermissions` call.
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:89]:
```suggestion
:::caution important
To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts), which automatically asks for permission to use `eth_accounts` by calling `wallet_requestPermissions` internally.
```

File: wallet/concepts/wallet-api.md [hunk 16]
```diff
@@ -61,23 +61,43 @@ Make sure to handle errors for every call to
...
+See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
+:::
+
+- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
+- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
+  
+:::info note
+For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](/snaps/learn/about-snaps/apis/#custom-json-rpc-apis) page.
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:97]: Maybe copy the example referenced here to the [Request permissions from a dapp](https://docs.metamask.io/snaps/how-to/request-permissions/#request-permissions-from-a-dapp) Snaps section and link there instead.

Comment (joaniefromtheblock) [reply wallet/concepts/wallet-api.md:95]: The entire section or just related to the getting the permissions? Am I duplicating the content for the custom JSON-RPC APIs?

**Comment by joaniefromtheblock** [comment]
@alexandratran updated, based on the suggestions. Does this look better?

File: snaps/how-to/request-permissions.md [hunk 17]
```diff
@@ -57,13 +57,31 @@ See the [`eth_accounts` dynamic permission](../reference/permissions.md#eth_acco
 Dapps that communicate with Snaps must request permission to do so by calling the
```
Comment (alexandratran) [inline snaps/how-to/request-permissions.md:57]:
```suggestion
Dapps that communicate with Snaps using [`wallet_snap`](/wallet/reference/wallet_snap) and [`wallet_invokeSnap`](/wallet/reference/wallet_invokesnap) must request permission to do so by calling the
```

File: snaps/how-to/request-permissions.md [hunk 18]
```diff
@@ -57,13 +57,31 @@ See the [`eth_accounts` dynamic permission](../reference/permissions.md#eth_acco
...
 [`wallet_requestSnaps`](/wallet/reference/wallet_requestsnaps) MetaMask JSON-RPC API method.
 
+:::info note
+Restricted [JSON-RPC](/wallet/reference/json-rpc-api) methods [`wallet_snap`](/wallet/reference/wallet_snap) 
+and [`wallet_invokeSnap`](/wallet/reference/wallet_invokesnap) require obtaining permission 
+through [`wallet_requestSnaps`](/wallet/reference/wallet_requestSnaps) before use.
+:::
+
```
Comment (alexandratran) [inline snaps/how-to/request-permissions.md:65]:
```suggestion
```

File: snaps/how-to/request-permissions.md [hunk 19]
```diff
@@ -57,13 +57,31 @@ See the [`eth_accounts` dynamic permission](../reference/permissions.md#eth_acco
...
 For example, to request permission to connect to the `hello-snap` Snap:
```
Comment (alexandratran) [inline snaps/how-to/request-permissions.md:66]:
```suggestion
The following example calls `wallet_requestSnaps` to request permission to connect to the `hello-snap` Snap, then calls `wallet_invokeSnap` to invoke the `hello` JSON-RPC method exposed by the Snap:
```

File: snaps/how-to/request-permissions.md [hunk 20]
```diff
@@ -57,13 +57,31 @@ See the [`eth_accounts` dynamic permission](../reference/permissions.md#eth_acco
...
+// Invoke the 'hello' JSON-RPC method exposed by the Snap.
+const response = await window.ethereum.request({
+  method: 'wallet_invokeSnap',
+  params: { snapId: 'npm:hello-snap', request: { method: 'hello' } },
+});
+
+console.log(response); // 'world!'
 ```
+
+See [About the Snaps APIs](/snaps/learn/about-snaps/apis/#custom-json-rpc-apis) for more information on implementing custom JSON-RPC APIs to communicate with dapps and other Snaps.
```
Comment (alexandratran) [inline snaps/how-to/request-permissions.md:87]:
```suggestion
:::note
Learn more about implementing [custom JSON-RPC APIs](../learn/about-snaps/apis.md#custom-json-rpc-apis) in a Snap.
:::
```

File: wallet/concepts/wallet-api.md [hunk 21]
```diff
@@ -61,23 +61,41 @@ Make sure to handle errors for every call to
...
+:::caution important
+To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts), which automatically asks for permission to use `eth_accounts` by calling `wallet_requestPermissions` internally.
+See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
+:::
+
+- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
+- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
+  
+:::info note
+For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the [requesting permissions from a dapp](/snaps/how-to/request-permissions/#request-permissions-from-a-dapp) example.
```
Comment (alexandratran) [inline wallet/concepts/wallet-api.md:95]:
```suggestion
For more information on using `wallet_snap` and `wallet_invokeSnap`, see the [how to request Snap permissions from a dapp](/snaps/how-to/request-permissions/#request-permissions-from-a-dapp).
```

**Review by alexandratran — APPROVED** [review]
LGTM once final suggestions are committed.
